documentation/internal/Assignment-Process.md

```diff
@@ -0,0 +1,57 @@
+![Review Assignment](images/Review-Assignment.png)
+
+First of all, some bulletpoint (and rules) to help clarify the diagram
+
+- One review assignment should exist per reviewer/application/stage/level
+- Review assignment can have the following statuses
+  - `Available` -> Can be assigned by assigner
+  - `Self-assigned by another` -> Was available for self assignment (but self assigned by someone else)
+  - `Assigned` -> At least one review_question_assignment was linked to review_assignment
+  - `Available for self-assignment` -> Can be self assigned
+- Review assignments can be restricted by section/s (template_section_restriction, which is an array of section codes user can review, value can be blank to allow review of all sections)
+- Review assignments of status `Available or Assigned` can be assigned by the assigner linked through the `review_assignment_assigner_join` table
+- `Assigned` status indicates that the reviewer is able to create a review (or one already exists)
+- Review assignment should only be editable by a reviewer if it's `Available for self-assignment`, otherwise by linked `Assigner`. Review assignment should not be editable once stage of review assignment is different to the current application stage.
+- `is_last_level` field is a helper field for front end (to avoid front ending needing to figure if if review is last level for given stage)
+
+## Difference between self-assignment and assignment
+
+The idea behind self assignment originated with requirement for users (who are not themselves assigners) to be able to 'choose' or pick which application they will be reviewing. Currently self assignment is not restricted by template_section_restrictions.
+
+## The Act of Assigning - Web App
+
+From front end perspective, this involves creating a mutation, which adds review_question_assignments to review_assignment, and changes status of review_assignment to `Assigned`. For self assignment, trigger is set to `ON_REVIEW_SELF_ASSIGN`.
+
+Front end doesn't generate review_assignment records, this happens on the server.
+
+## Server Side
+
+To reduce front end logic, server is responsible for generating review_assignment records, based on user permissions. This is done by an action `generateReviewAssignments` which is configured to fire when application or review changes status.
+
+Basically, when application is submitted, we want to look at all of the permissions linked to application template of type Review, where template permission configurations match current level and stage, review_assignment records are generated for users who are linked to that template permission (through permission name), and any extra configurations (template section restrictions, canSelfAssign, etc..) are populated for new review_assignment record.
+
+When reviews are submitted, we do similar thing but for level + 1. Typical scenario is a consolidator being informed that they can self assign a consolidation when lower level reviewer submits a review.
```

Comment:
Moved things a lot on this suggestion. Basically deals first with what was in "Server side" with title "Pre-assignment process" and list what happens on the front end under the same title "Web app" but now with sub-titles of "Self-assignment" and "Assigning to others".

```suggestion

## Pre-assignment process

To reduce front-end logic, server is responsible for generating `review_assignment` records, based on user permissions. This is done by an action called `generateReviewAssignments` which is configured to fire when application or review changes status.
e.g.: Application is Submitted, Review is approved and moves to new Stage.

### Application submission
The server logic looks at all of the permissions linked to application template of type `review`, where template permission configurations match current `level` and `stage`. It creates `review_assignment` records for users who are linked to that template permission (through permission name), and any extra configurations (`template_section_restrictions`, `canSelfAssign`, etc..) are populated for new review_assignment record.

### Review submission
The server logic does similar thing as what is done when Application is submitted, but for level + 1. Typical scenario is a consolidator being informed that they can self assign a consolidation when lower level reviewer submits a review.

## Assigning process
Available for either a Reviewer that can do **self-assignment** or an Assigner with permission to **assign** applications of a certain template type on a `level` and `stage` (settings defined on the `template_permission` of each Template).

 To **assign** or **self-assignment** the Reviewer of sections of one application there are option available on the Applications list of on the Application home page.  

### Web App

#### Self-assignment
A GraphQL mutation is created the front-end when the user clicks on the **Self-assign** for one section.
What happens as a result of this mutation is: 
1. `review_assignment.trigger` is set to ON_REVIEW_SELF_ASSIGN.
2. Changes `review_assignment.status` to **Assigned**. 
3. Adds `review_question_assignment` (for each question of section assigned) to `review_assignment`
4. Other updates done on the server-side:
  - Update other `review_assignment.status` from **Available for self-assignment** to **Self-assigned by another**
  - Assign other sections to the same reviewer - to avoid misleading half-review.

#### Assigning to others
A GraphQL mutation  is created the front-end when the user selects one Reviewer on the drop-down list to **assign** for one section. What happens as a result of this mutation is: 
1. `review_assignment.trigger` is set to ON_REVIEW_ASSIGN.
2. Changes status of `review_assignment` to **Assigned**.
3. Adds `review_question_assignment` (for each question on the section assigned) to `review_assignment`
4. Other updates done on the server-side:
  - Update other `review_assignment.status` from **Available** to **Locked**
```